### PR TITLE
Fix 3D SIV Delay & Median Move

### PR DESCRIPTION
**Pull request by amin1377**
- Add the scaled delay from the "koios_3d" benchmark into the 3D SIV-like architecture. The inter-die delay is calculated by multiplying the delay of the L4 segment by the ratio of the inter-die delay to the L4 delay of the Koios_3D architecture.
- Fix median/weighted median move to get the median layer of bounding boxes instead of pins (as it is the case for choosing x and y locations)

### CODE REVIEW

**Comment by amin1377** [comment]
2D Titan: 
Screenshot from 2024-03-04 16-08-26

**Comment by vaughnbetz** [comment]
Add 3D results.

File: vpr/src/place/centroid_move_generator.cpp [hunk 1]
```diff
@@ -39,8 +39,8 @@ e_create_move CentroidMoveGenerator::propose_move(t_pl_blocks_to_be_moved& block
     calculate_centroid_loc(b_from, false, centroid, nullptr);
 
     // Centroid location is not necessarily a valid location, and the downstream location expect a valid
-    // layer for "to" location. So if the layer is not valid, we set it to the same layer as from loc.
-    to.layer = (centroid.layer < 0) ? from.layer : centroid.layer;
+    // layer for the centroid location. So if the layer is not valid, we set it to the same layer as from loc.
```
Comment (vaughnbetz) [inline vpr/src/place/centroid_move_generator.cpp:42]: expect -> expects

File: vpr/src/place/median_move_generator.cpp [hunk 2]
```diff
@@ -101,7 +112,20 @@ e_create_move MedianMoveGenerator::propose_move(t_pl_blocks_to_be_moved& blocks_
...
             }
 
-            if (!get_bb_incrementally(net_id, coords, xold, yold, xnew, ynew)) {
+            if (net_bb_coords.layer_min == layer_old) {
+                layer_new = net_bb_coords.layer_max;
+            } else {
+                layer_new = net_bb_coords.layer_min;
+            }
+
+            if (!get_bb_incrementally(net_id,
```
Comment (vaughnbetz) [inline vpr/src/place/median_move_generator.cpp:121]: Make a comment of why this works.

File: vpr/src/place/place.cpp [hunk 3]
```diff
@@ -2902,11 +2924,18 @@ static void get_bb_from_scratch(ClusterNetId net_id,
     coords.xmax = xmax;
     coords.ymin = ymin;
     coords.ymax = ymax;
+    coords.layer_min = layer_min;
+    coords.layer_max = layer_max;
+    VTR_ASSERT(layer_min >= 0 && layer_min < device_ctx.grid.get_num_layers());
```
Comment (vaughnbetz) [inline vpr/src/place/place.cpp:2929]: might make this a VTR_ASSERT_DEBUG (reasonably hot function).

File: vpr/src/place/placer_context.h [hunk 4]
```diff
@@ -113,6 +113,7 @@ struct PlacerMoveContext : public Context {
     // These vectors will grow up with the net size as it is mostly used to save coords of the net pins or net bb edges
     std::vector<int> X_coord;
```
Comment (vaughnbetz) [inline vpr/src/place/placer_context.h:114]: Could experiment with the speed of a array of structs (t_pos I think).

Comment (vaughnbetz) [reply vpr/src/place/placer_context.h:114]: OK, scratch that as you have to sort them.  But comment that (reason why we use independent vectors).

File: vtr_flow/arch/multi_die/stratixiv_3d/3d_full_OPIN_inter_die_stratixiv_arch.timing.xml [hunk 5]
```diff
@@ -5115,9 +5115,11 @@
               while keeping the switch delay a reasonable (see comment in <segmentlist>) portion of the overall delay.
        -->
     <switch type="mux" name="seg4_driver" R="450" Cin="0.60e-15" Cout="4.82e-15" Tdel="59e-12" mux_trans_size="2.630740" buf_size="27.645901"/>
+    <switch type="mux" name="seg4_inter_layer_driver" R="0.0" Cin="0.0" Cout="0.0" Tdel="138.82e-12" mux_trans_size="2.630740" buf_size="27.645901"/>
```
Comment (vaughnbetz) [inline vtr_flow/arch/multi_die/stratixiv_3d/3d_full_OPIN_inter_die_stratixiv_arch.timing.xml:5118]: Could comment why this delay is what it is (reasoning behind taking it from Koios).

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
A few changes suggested, and please get the 3D data.

**Comment by vaughnbetz** [comment]
@amin1377 : is this one ready to merge?

**Comment by amin1377** [comment]
@vaughnbetz: I've just added more comments and rebased it with the master branch. Once the CI tests pass successfully, it'll be ready for merging.

**Pull request merged**